### ISSUE TITLE
Test the model override form

### ISSUE THREAD

**Issue by davidszkiba** [Wunderbyte-GmbH/moodle-local_catquiz#410]
This is about testing the following form: 

![Screenshot 2024-02-13 at 09 30 13](https://github.com/Wunderbyte-GmbH/moodle-local_catquiz/assets/78353570/3d18a1bf-68f9-4b8e-8e67-caef56a43965)

If you navigate to the catquiz manager via the button in the main navigation menu, then select the "Questions" tab and there select a scale, you should see a table that lists all questions assigned to the selected CAT scale.
In that question table, you can click on the cogwheel icon in the Action column to get more information about that question. On that page, there is this form that lets you edit which item parameter is connected to the selected question (there is one item param for each model: one for 1PL rasch, one for 2PL raschbirnbaum and one for 3PL mixedraschbirnbaum). You see the form once you click "Edit"

So the information about an item is displayed in two modes: as static text and, when clicking on the "Edit" button, as form where the values can be edited.

Please test the following things

- [x] When in edit mode, if the selected status is "not yet calculated", the parameter fields (difficulty, guessing, ...)  should be hidden.
- [x] When in edit mode, if the selected status is "calculated", the parameter fields should be visible but disabled.
- [ ] Trying to save two or more items with status manually confirmed should show a validation error.
- [ ] Values are saved correctly (e.g. after page reload still visible) and also visible after clicking "Save", i.e. displayed as static text.
- [ ] When navigating back to the Questions tab, (this has to be done via a back button right now), the updated values are also listed in the table.

Note: there is another card on that page that shows the highest selected status among all models and the name of the corresponding model. This will show updated values only after a page reload.

The status order is as follows:
manually confirmed > manually updated > calculated > not yet calculated > manually excluded

We should also check that the information here is correct:

- [ ] After selecting a new model:
  Assume 1PL rasch is selected because it has status "manually confirmed". You click Edit, set the status for 1PL rasch to excluded and select "manuallly confirmed" for the 2PL raschbirnbaum model. Please check that after a page reload the "Status" card shows that the status is "manually confirmed" and the selected model is now 2PL raschbirnbaum.
- [ ] Check the same with other status combinations. Always the highest status should be shown in the "Status" card

**Comment by semteacher** [comment]
@davidszkiba 
1. In case of "true/false" questions - cannot open "mamage details" (cogwheel) page for question. Got error
```
Exception - Warning: Attempt to read property "status" on bool in [dirroot]/local/catquiz/classes/output/testitemdashboard.php on line 228
Stack trace:
line 157 of /lib/behat/lib.php: Exception thrown
line 228 of /local/catquiz/classes/output/testitemdashboard.php: call to behat_error_handler()
line 316 of /local/catquiz/classes/output/testitemdashboard.php: call to local_catquiz\output\testitemdashboard->get_itemstatus()
line 187 of /local/catquiz/classes/output/catscalemanager/managecatscaledashboard.php: call to local_catquiz\output\testitemdashboard->return_as_array()
line 66 of /local/catquiz/manage_catscales.php: call to local_catquiz\output\catscalemanager\managecatscaledashboard->__construct()
```
2. Multichoice questions, 2PL Rasch-Birnbaum model - cannot save changes from "Manually updated" to "Calculated". Got error:
```
Exception - Warning: Undefined property: stdClass::$override_mixedraschbirnbaum in [dirroot]/local/catquiz/classes/form/item_model_override_selector.php on line 363
Stack trace:
* line 157 of /lib/behat/lib.php: Exception thrown
* line 363 of /local/catquiz/classes/form/item_model_override_selector.php: call to behat_error_handler()
* line 207 of /local/catquiz/classes/form/item_model_override_selector.php: call to local_catquiz\form\item_model_override_selector->generate_model_fields()
* line 75 of /lib/form/classes/external/dynamic_form.php: call to local_catquiz\form\item_model_override_selector->process_dynamic_submission()
* line ? of unknownfile: call to core_form\external\dynamic_form::execute()
* line 253 of /lib/external/classes/external_api.php: call to call_user_func_array()
* line 83 of /lib/ajax/service.php: call to core_external\external_api::call_external_function()
```
In fact - got same type of errors for any attempt to save changes

**Comment by davidszkiba** [comment]
@semteacher thanks, I take a look

**Comment by eynimeni** [comment]
@semteacher we looked into 1) and were wondering, where you got the true/false question from? did you import the corresponding itemparams with the catmanager importer (csv) ?

**Comment by semteacher** [comment]
@eynimeni I have used [this](https://github.com/Wunderbyte-GmbH/moodle-local_catquiz/blob/main/tests/fixtures/mathematik2scales.xml) fixture file with 3 true/false questions for basic quick tests for a long time. 
If plugin does not supoort specific question types - it has to be explicitly noted. Or even prohibited to add / import such types of questions - to not confuse users. That is my opinion and reasons I reported above issue

**Comment by eynimeni** [comment]
ok thank you. we will find a way to treat questions that don't have itemparams assigned. thank you.

**Comment by semteacher** [comment]
@davidszkiba The issues described in [this comment ](https://github.com/Wunderbyte-GmbH/moodle-local_catquiz/issues/410#issuecomment-2056887279) had not been resolved yet. And it appears for any types of questions.
I have added steps to the behat to test "Manually confirmed" case and got one of that [errors ](https://screenrec.com/share/lzBYPgIV71) Moreover - I am receiving exactly the same warnings when tying to perform manual editing after interuption of automated test. Those warnings might not appears during maual tests if error level is too high - but they always breaks test because errorlevel is maximal in this case. So it is necessary to fix it. I can try but you definitely know object model of catquiz better. Could you take a look into it?

**Comment by davidszkiba** [comment]
@semteacher Ok thank you. You're right, the underlying issue should be fixed. And yes, I'll take a look.

**Comment by eynimeni** [comment]
Update unter Berücksichtigung der neuen Column in den Itemparametern

**Comment by davidszkiba** [comment]
I will update the form according to the latest changes to the DB structure:

Each item in the local_catquiz_items table has a field `activeparamid`, that points to the item params that are used to represent that item.

- [ ] Add a form select to choose the active model (itemparam).
- [ ] When a model is selected, an edit form will be rendered to allow overwriting the parameters.
- [ ] If there is no item parameter yet for the chosen model, the button says "Add", otherwise it says "Update"

**Comment by davidszkiba** [comment]
-> fertig machen, nachdem polytome items in der DB gespeichert werden können

**Comment by davidszkiba** [comment]
@ralferlebach 

So sieht im Moment die Übersicht aus:

![Image](https://github.com/user-attachments/assets/14f9c785-c989-4ce5-a685-e1d8e3854f80)

Wenn man ein Item bearbeitet, sieht es so aus:

![Image](https://github.com/user-attachments/assets/b7b573f2-ba69-4dc5-9025-5ee7073d7359)

Ich hab mal kurz mit Georg darüber gesprochen, und der Vorschlag wäre, dass wir:

- Weiter oben ein zusätzliches Dropdown Element hinzufügen, in dem man das aktive Modell auswählen kann.
- Zusätzlich gibt es neben den Modellen einen Radio button, über den man ebenfalls das aktive Modell setzen kann. Die Radio Buttons sind mit dem Dropdown verknüpft: wenn man im Dropdown das aktive Element ändert, ändern sich auch die Radio Buttons - und umgekehrt.

Es muss noch berücksichtigt werden, was mit dem Status passiert, wenn ein Modell ausgewählt wird.
Im Moment gibt es folgende Status Werte:

- Not yet calculated
- Calculated
- Manually excluded
- Manually updated
- Manually confirmed

Soll ein Modell mit Status "Manually excluded" überhaupt ausgewählt werden können? Wenn ja, ändert sich der Status automatisch auf "Manually confirmed"? Was passiert dann mit anderen Modellen, die Status "Manually confirmed" haben - behalten die diesen Status oder soll er dort herabgesetzt werden?

**Comment by eynimeni** [comment]
- [ ] Mittlere Karte raus, 
- [ ] rechte KArte möglichst 2/3 breit
- [ ] Likelyhood heißt nun "Ausgewähltes Modell"
- [ ] Neue Überschrift H3 Modellname
- [ ] Status (normale Größe)
- [ ] Grafik wie gehabt
- [ ] bei Grafik Y-Achse fix von 0 weg starten
